// @arcgis/eslint-config/dist/config/index.js
import {
  webgis_default
} from "../chunk-G2LC3KBE.js";
import "../chunk-FVTAJXPI.js";

// src/config/index.ts
import eslint from "@eslint/js";
import tsEslint from "typescript-eslint";
import restrictedGlobals from "confusing-browser-globals";
import globals from "globals";
var config_default = [
  {
    ignores: ["**/www", "**/dist", "**/assets", "**/coverage", "**/.docs"]
  },
  eslint.configs.recommended,
  ...tsEslint.configs.strictTypeChecked,
  ...tsEslint.configs.stylisticTypeChecked,
  {
    ...webgis_default.configs.recommended,
    files: ["**/*.ts", "**/*.tsx"]
  },
  {
    files: ["**/*.ts", "**/*.tsx"],
    linterOptions: {
      reportUnusedDisableDirectives: "error"
    },
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node,
        ...globals.jest
      },
      ecmaVersion: "latest",
      sourceType: "module"
    },
    rules: {
      //#region Disable rules redundant with Typescript
      /**
       * The following rules are disabled because they are redundant with
       * TypeScript. No need to needlessly slow down ESLint and clutter the
       * error messages.
       */
      // TypeScript reports this
      "no-unused-labels": "off",
      "@typescript-eslint/no-unused-vars": [
        "off",
        // This configuration matches TypeScript
        {
          // Allow args and vars that begin with _
          // It is sometimes useful to define an unused variable for readability
          // (i.e when doing array destructuring).
          // TypeScript ignores unused variables starting with _ too
          argsIgnorePattern: "^_",
          varsIgnorePattern: "^(_|(h|Fragment)$)"
        }
      ],
      // Redundant with TypeScript, but has autofix, so enabled for convenience
      "no-unsafe-negation": "warn",
      // TypeScript reports this
      "no-unused-private-class-members": "off",
      // TypeScript detects re-declarations like this
      "no-case-declarations": "off",
      // Narrow use case. Even less with TypeScript
      "no-ex-assign": "off",
      // Redundant with no-restricted-globals and with TypeScript
      "no-global-assign": "off",
      // Redundant with prefer-regex-literals, which enforces regex literal
      // syntax, which is checked by TypeScript
      "no-invalid-regexp": "off",
      // TypeScript reports this as syntax error
      "no-nonoctal-decimal-escape": "off",
      // TypeScript reports this as syntax error
      "no-octal": "off",
      // TypeScript reports this as type error
      "no-unsafe-optional-chaining": "off",
      // TypeScript emits an error for with() statements
      "no-with": "off",
      // TypeScript reports this as a type error
      "valid-typeof": "off",
      // TypeScript reports this
      "@typescript-eslint/prefer-namespace-keyword": "off",
      //#endregion
      //#region Disable rules Redundant with Prettier
      /**
       * The following rules are redundant with Prettier
       */
      "no-unexpected-multiline": "off",
      // Outside of comments, this is redundant with Prettier.
      // In comments, it is not an issue since comments are stripped in minified
      // builds
      "no-irregular-whitespace": "off",
      //#endregion
      //#region Disable obsolete rules
      /**
       * Obsolete rules (rules that enforced constraints that are no longer
       * present in the language, or that are no longer relevant)
       */
      // Namespaces are needed for defining custom elements
      "@typescript-eslint/no-namespace": "off",
      // A lot of false positives when dealing with Arcade
      "no-template-curly-in-string": "off",
      // We don't use TSLint
      "@typescript-eslint/ban-tslint-comment": "off",
      //#endregion
      //#region Disable rules that were moved to extra config
      /**
       *
       * The following default rules are disabled in favor of moving them into the
       * @arcgis/eslint-config/extra due to huge number of false positives or
       * performance issues.
       *
       */
      // Disabled as per https://devtopia.esri.com/WebGIS/arcgis-web-components/discussions/446#discussioncomment-3166
      "@typescript-eslint/no-unsafe-return": "off",
      /*
       * Lot's of false positives on pure functions that don't use
       * "this". For example:
       * const { whenOnce } = await importCoreReactiveUtils();
       */
      "@typescript-eslint/unbound-method": "off",
      /**
       * False positive when constructor exists to improve typing/narrow down the
       * type
       */
      "@typescript-eslint/no-useless-constructor": "off",
      //#endregion
      //#region Detect possible bugs
      /**
       *
       * The following rules report what is a possible bug. If you are sure that the
       * behavior is intentional, you can disable the rule for that line. However,
       * consider refactoring to less confusing syntax instead
       *
       */
      // Might be intentional, so only a warning
      "no-await-in-loop": "warn",
      "no-constructor-return": "warn",
      "no-fallthrough": ["error", { reportUnusedFallthroughComment: true }],
      // This mis-fires for anonymous functions when not wrapped in {}
      "no-promise-executor-return": "off",
      "no-self-compare": "error",
      "no-unreachable-loop": "error",
      // This rule can provide optional optimizations. It might be redundant
      // with TypeScript in some cases, but it provides a less-confusing error
      // message (for example in .reduce())
      "array-callback-return": "warn",
      // Good rule to have, but many false positives, so only a warning
      "require-atomic-updates": "warn",
      "grouped-accessor-pairs": "error",
      "no-extra-bind": "warn",
      "no-multi-assign": "error",
      // Not an error as some external APIs require calling "new" even if we don't
      // care about the retune time
      "no-new": "warn",
      "no-useless-computed-key": ["warn", { enforceForClassMembers: true }],
      "no-useless-concat": "error",
      "no-unneeded-ternary": "warn",
      "no-useless-call": "error",
      "no-useless-rename": "warn",
      // Useless returns are sometimes used for better readability
      // i.e "default: return;"
      "no-useless-return": "off",
      // Prevent accidental usages of global when forgot to declare a
      // variable. If you actually mean to use them, prepend with "globalThis."
      "no-restricted-globals": [
        "error",
        .../* @__PURE__ */ new Set([
          ...restrictedGlobals,
          "event",
          "name",
          "closed",
          "i",
          "index",
          "length",
          "parent",
          "self",
          "status",
          "stop",
          "toolbar",
          "top",
          // use Number.POSITIVE_INFINITY instead
          "Infinity",
          // use Number.NaN instead
          "NaN",
          // use Number.isNaN instead
          "isNaN",
          // Use Number.isFinite instead
          "isFinite",
          // use Number.parseFloat instead
          "parseFloat",
          // use Number.parseInt instead
          "parseInt",
          // Use Object.keys instead
          "keys"
        ])
      ],
      "@typescript-eslint/no-useless-empty-export": "warn",
      // Consider using String.prototype.localeCompare()
      "@typescript-eslint/require-array-sort-compare": "warn",
      "no-object-constructor": "error",
      /**
       * Something as simple as this cannot be typed without a TypeScript error
       * without using void in place of undefined:
       * const a = (callback: ()=>string | undefined) => {};
       * a(()=>{ console.log('1'); })
       */
      "@typescript-eslint/no-invalid-void-type": "off",
      "@typescript-eslint/no-empty-object-type": ["warn", { allowInterfaces: "with-single-extends" }],
      "@typescript-eslint/no-floating-promises": ["warn", { ignoreVoid: true }],
      "@typescript-eslint/restrict-template-expressions": [
        "warn",
        {
          allowNumber: true
        }
      ],
      "@typescript-eslint/restrict-plus-operands": ["warn", { allowNumberAndString: true }],
      // We often actually have control characters in our regexes
      "no-control-regex": "off",
      // Functions that deal with JSON.stringify/localStorage may have a type
      // parameter as a more readable alternative to a type assertion.
      // Also, this rule is quite complicated/expensive.
      "@typescript-eslint/no-unnecessary-type-parameters": "off",
      // Disabling until this is fixed:
      // https://github.com/typescript-eslint/typescript-eslint/issues/9057
      "@typescript-eslint/use-unknown-in-catch-callback-variable": "off",
      /**
       *
       * Disallow deprecated and discouraged language syntax
       *
       * Redundant rules are not enabled (i.e since we don't allow using labels
       * at all, rules that detect incorrect uses of labels aren't enabled)
       *
       */
      "no-restricted-syntax": [
        /**
         * To help you build an AST selector, you can use https://astexplorer.net/
         * (be sure to set parser to "@typescript-eslint/parser"). Or, you can use
         * https://typescript-eslint.io/play/
         *
         * Documentation of AST selectors and example selectors:
         * https://eslint.org/docs/latest/extend/selectors
         */
        "error",
        {
          selector: "ForInStatement",
          message: "for..in loops iterate over the entire prototype chain, which is virtually never what you want. Instead use for of/.forEach() with Object.keys/values/entries.forEach()"
        },
        {
          selector: "LabeledStatement",
          message: "Labels are not allowed"
        },
        {
          selector: 'ImportSpecifier[imported.name="createElement"][local.name!="h"]',
          message: 'Replace with "createElement as h" (this allows using a single tsconfig.json both for Lumina and Storybook React)'
        }
      ],
      // We forbid for-in completely, so no need to detect incorrect usages
      "@typescript-eslint/no-for-in-array": "off",
      "no-alert": "error",
      "no-caller": "error",
      // TODO: re-enable once https://devtopia.esri.com/WebGIS/arcgis-web-components/issues/1519 is fixed
      // "no-console": 'error',
      "no-eval": "error",
      "no-extend-native": "error",
      // This creates less readable code
      "no-return-assign": "error",
      "no-script-url": "error",
      // This creates less readable code
      "no-sequences": "error",
      "prefer-arrow-callback": ["warn", { allowNamedFunctions: true }],
      "prefer-numeric-literals": "warn",
      "prefer-object-spread": "warn",
      "strict": "warn",
      "prefer-template": "warn",
      "prefer-object-has-own": "warn",
      // Lots of false positives
      "@typescript-eslint/no-base-to-string": "off",
      "@typescript-eslint/no-confusing-void-expression": [
        "error",
        {
          ignoreArrowShorthand: true,
          ignoreVoidOperator: true
        }
      ],
      // Conflicts with @typescript-eslint/no-confusing-void-expression
      // See https://github.com/typescript-eslint/typescript-eslint/issues/4507
      "@typescript-eslint/no-meaningless-void-operator": "off",
      "@typescript-eslint/no-empty-function": [
        // There can be use cases for this (when providing a callback is required
        // but you don't want to do anything)
        "warn",
        {
          allow: ["decoratedFunctions", "overrideMethods"]
        }
      ],
      "no-loss-of-precision": "error",
      "@typescript-eslint/switch-exhaustiveness-check": "error",
      "default-case-last": "error",
      "func-style": ["error", "declaration", { allowArrowFunctions: true }],
      "no-new-func": "error",
      "no-new-wrappers": "error",
      "@typescript-eslint/no-misused-promises": [
        "warn",
        {
          // Redundant with TypeScript
          checksSpreads: false,
          checksVoidReturn: {
            arguments: true,
            // Allow for patterns like ref={focusElement}
            attributes: false,
            inheritedMethods: true,
            properties: true,
            returns: true,
            variables: true
          }
        }
      ],
      //#endregion
      //#region Report consistency issues
      // TODO: make these more or less strict as needed
      /**
       * Consistency and clean code rules
       */
      "arrow-body-style": "warn",
      // This rule is a good idea, but not autofixable, so is only a warning
      // TODO: this does not support type-only imports. use import/no-duplicates instead
      "no-duplicate-imports": "off",
      // We will enforce strict curly. Rational:
      // - Same as core api
      // - Better for debugging and breakpoints
      "curly": ["warn", "all"],
      "eqeqeq": ["error", "always", { null: "ignore" }],
      "func-names": ["warn", "as-needed"],
      // Rationale: https://devtopia.esri.com/WebGIS/arcgis-web-components/issues/289
      // TODO: replace with unicorn/prevent-abbreviations
      "id-denylist": ["error", "elt", "elem", "evt", "ev", "cb"],
      "complexity": ["warn", { max: 30 }],
      "new-cap": [
        "warn",
        {
          capIsNewExceptions: [
            // Exclude Stencil decorators
            "Component",
            "Element",
            "Event",
            "AttachInternals",
            "Listen",
            "Method",
            "Prop",
            "State",
            "Watch",
            // Exclude Angular decorator
            "NgModule"
          ]
        }
      ],
      // TODO: discuss (this avoids longer lines, improves readability)
      // "no-inline-comments": "error",
      // TODO: discuss (similar to the rule above, might be redundant).
      // "@stylistic/line-comment-position": "error",
      // Explicit boolean cast makes code more readable and catches potential bugs
      "no-extra-boolean-cast": "off",
      /**
       * Code like this is useful for readability:
       * ```ts
       * const component = this;
       * ```
       */
      "@typescript-eslint/no-this-alias": "off",
      // ENUM values like "1<<0" ... "1<<9" are very useful
      "@typescript-eslint/prefer-literal-enum-member": "off",
      "no-magic-numbers": "off",
      "@typescript-eslint/no-magic-numbers": [
        // Error on code like:
        //   switch (code) { case 9: // '\t' ...
        // refactor to:
        //   export const charCodes = { tab: 9 };
        //   switch (code) { case charCodes.tab: ...
        //
        // Benefits:
        // - charCodes can be exported and reused in the codebase
        // - no need for '\t' comment anymore
        //   - comment can be wrong or get out of data
        // - easy to find all usages of charCodes.tab (harder to search for usages
        //   of 9)
        //
        // Similarly, you can define a dictionary of all HTTP status codes and
        // other special numbers. Example:
        // https://github.com/specify/specify7/blob/xml-editor/specifyweb/frontend/js_src/lib/utils/ajax/definitions.ts
        // And now it's trivial to see all the places in the code that handle a
        // particular http status code.
        // Similarly, in that file further down you can see how by using
        // [Http.Conflict] rather than 409:, the code is more self-documenting
        "warn",
        {
          ignore: [
            // Don't know why these aren't ignored by default
            -1,
            0,
            1,
            2
          ],
          ignoreDefaultValues: true,
          ignoreClassFieldInitialValues: true,
          detectObjects: false,
          ignoreEnums: true,
          ignoreNumericLiteralTypes: true,
          ignoreReadonlyClassProperties: true,
          ignoreTypeIndexes: true
        }
      ],
      "no-warning-comments": [
        // TODO: consider forbidding TODO in favor of BUG, HACK, REFACTOR, FEATURE,
        //   LOW, TEST, PERF. Then can configure IDE to assign different colors to
        //   each of those
        "error",
        {
          // By emitting error for these TODOs, you can use them for things you
          // want to do in the current commit (since they will be reported as
          // an error by pre-commit hook). Convenient for not-forgetting to do
          // something
          terms: ["ME"],
          location: "anywhere"
        }
      ],
      // Autofixable, so warn to be less disruptive
      "object-shorthand": "warn",
      // This kind of variable declaration is harder to refactor:
      // - That style is hard to write a regex for
      // - That style makes it harder to refactor code (you can't just copy paste
      //   a line and modify it, because you have to make sure that only first
      //   line has "const", and only last line has ";") -> that also makes git's
      //   diff larger.
      // - You can't easily convert a variable between let and const or back
      //   without moving the line - makes refactoring harder, and makes git's
      //   diff larger
      "one-var": ["warn", "never"],
      "prefer-regex-literals": ["error", { disallowRedundantWrapping: true }],
      // "u" flag in regex does two important things:
      // - support unicode surrogate paris (properly parse glyphs that are more
      //   than one code point - i.e emojis)
      // - enable strict mode for regex - i.e
      //   - this is valid but ambiguous - /{a}/
      //   - this throws syntax error - /{a}/u
      //   - this is good and not ambiguous - /\{a\}/u
      //
      // The one case when you don't want "u" is when RegExp is coming from the
      // user and you want to be more lenient with parsing (u enables stricter
      // Regex)
      "require-unicode-regexp": "error",
      // TODO: discuss (or maybe use simple-import-sort)
      // 'sort-imports': 'error',
      "symbol-description": "warn",
      // Autofixable, so only a warning
      "yoda": "warn",
      "unicode-bom": "warn",
      // Math.pow is more readable than **
      "prefer-exponentiation-operator": "off",
      "@typescript-eslint/ban-ts-comment": [
        "error",
        {
          "ts-expect-error": "allow-with-description",
          "ts-ignore": false,
          "ts-nocheck": "allow-with-description",
          "ts-check": true,
          "minimumDescriptionLength": 3
        }
      ],
      // Getters are slower than direct property access. And if library user has
      // TypeScript, overwriting readonly will be flagged as an error
      "@typescript-eslint/class-literal-property-style": "off",
      "@typescript-eslint/array-type": ["warn"],
      "@typescript-eslint/consistent-type-assertions": [
        // In some cases object literal type assertions are necessary
        "warn",
        {
          assertionStyle: "as",
          objectLiteralTypeAssertions: "allow-as-parameter"
        }
      ],
      "@typescript-eslint/consistent-type-exports": ["warn", { fixMixedExportsWithInlineTypeSpecifier: true }],
      "@typescript-eslint/consistent-type-imports": "warn",
      "default-param-last": "off",
      // `undefined` can also be used to trigger default value
      "@typescript-eslint/default-param-last": "off",
      "@typescript-eslint/explicit-function-return-type": [
        // Providing explicit type improves type-checking performance and
        // catches bugs (when actual return type is not what you are expecting,
        // or return type changed accidentally during incorrect refactoring).
        // It also improves code documentation
        // In some cases, explicit type would be very bloated/complicated (i.e in
        // case of mixins), thus you can silence this rule in those cases
        "error",
        {
          allowConciseArrowFunctionExpressionsStartingWithVoid: true
        }
      ],
      "max-params": "off",
      // Two spaces is more readable than " {2}"
      "no-regex-spaces": "off",
      "prefer-const": "warn",
      "no-var": "error",
      // Prefer passing parameters as an object instead to be less error prone
      // and make adding/removing properties require less refactoring
      "@typescript-eslint/max-params": ["warn", { max: 5 }],
      // "property" style is more type-safe then "method" (has covariance checks)
      "@typescript-eslint/method-signature-style": "warn",
      "@typescript-eslint/no-inferrable-types": ["warn"],
      "@typescript-eslint/non-nullable-type-assertion-style": ["warn"],
      "@typescript-eslint/dot-notation": "warn",
      "@typescript-eslint/prefer-function-type": ["warn"],
      "@typescript-eslint/prefer-optional-chain": ["warn"],
      "@typescript-eslint/prefer-string-starts-ends-with": ["warn"],
      "@typescript-eslint/no-array-constructor": ["warn"],
      "@typescript-eslint/no-duplicate-type-constituents": ["warn"],
      "@typescript-eslint/no-unnecessary-boolean-literal-compare": ["warn"],
      "@typescript-eslint/no-extra-non-null-assertion": ["warn"],
      "@typescript-eslint/no-unnecessary-type-arguments": ["warn"],
      "@typescript-eslint/no-unnecessary-type-assertion": ["warn"],
      "@typescript-eslint/prefer-includes": ["warn"],
      "@typescript-eslint/prefer-reduce-type-parameter": ["warn"],
      "@typescript-eslint/prefer-return-this-type": ["warn"],
      "@typescript-eslint/naming-convention": [
        "error",
        {
          /*
           * Use camelCase and PascalCase by default
           * Pascal case is allowed because, despite variable names
           * commonly starting with lower case, there are use cases for
           * capital letter:
           *  - functional components
           *  - React contexts
           *  - constructors/dynamically created classes/mixins
           *  - enum-like config objects
           */
          selector: "default",
          format: ["camelCase", "PascalCase"],
          // Denotes unused variables. Also may denote private members
          leadingUnderscore: "allow",
          trailingUnderscore: "forbid"
        },
        {
          // TypeLike must be in PascalCase
          selector: "typeLike",
          format: ["PascalCase"]
        },
        {
          // Relax restrictions for keys like `en-US` and etc.
          selector: "default",
          modifiers: ["requiresQuotes"],
          format: null
        },
        {
          // Enum members should be in upper case
          selector: "enumMember",
          format: ["UPPER_CASE"]
        }
      ],
      // TODO: turn this into an error. Discuss with the team why "any" defeats the purpose and benefits of TypeScript.
      // To get all benefits of TypeScript, you have to go all full-in. Use of any
      // may indicate poor code quality, or lack of understanding of TypeScript.
      // In extreme cases, "unknown" should be used instead of any. There are
      // virtually no correct use cases of any
      "@typescript-eslint/no-explicit-any": "warn",
      // There are cases where non-null assertions are useful as TypeScript fails
      // to detect the preceding non-null check. Example:
      // const key = 'key' as string; // Some dynamic key
      // const a: Record<string, string[] | undefined> = {};
      // a[key] ??= [];
      // a[key].push("a"); // False error
      //
      // Also, while you can explicitly check for non-null and throw an exception,
      // if you use !, the browser will automatically throw an exception for you:
      // if(someObject == null) throw new Error('someObject is not defined');
      // someObject.a();
      // is equivalent to:
      // someObject!.a();
      // (and browser's exception message may be more useful)
      //
      // More thoughts on this:
      // https://github.com/maxpatiiuk/dotfiles/blob/main/npm%2Feslint-config%2Feslint.config.js#L577-L602
      "@typescript-eslint/no-non-null-assertion": "off",
      "@typescript-eslint/no-require-imports": "warn",
      // Better documents code, and if exception occurs, results in a rejected
      // promise rather than an unhandled exception
      "@typescript-eslint/promise-function-async": "warn",
      // Not an error because:
      // If passing function as a prop, and the function is expected to
      // return a promise, making function async is cleaner than
      // returning Promise.resolve()
      "@typescript-eslint/require-await": "warn",
      // This improves stack traces, improves performance, and makes code more
      // explicit
      "@typescript-eslint/return-await": ["warn", "always"],
      // TODO: replace with perfectionist/sort-intersection-types and perfectionist/sort-union-types rules.
      // Autofixable, so only a warning to be less disruptive
      "@typescript-eslint/sort-type-constituents": "warn",
      "@typescript-eslint/prefer-nullish-coalescing": [
        "warn",
        {
          ignorePrimitives: { string: true }
        }
      ],
      // The string.match() is more intuitive than the RegExp.exec()
      "@typescript-eslint/prefer-regexp-exec": "off",
      // Both types and interfaces have uses in the edge cases. In the common
      // cases it often does not matter.
      // See alsohttps://www.youtube.com/watch?v=zM9UPcIyyhQ
      "@typescript-eslint/consistent-type-definitions": ["off", "type"],
      // This is error by default. Turning into warning because it falsely
      // reports when component tries to emit an event that is marked as
      // deprecated.
      "@typescript-eslint/no-deprecated": "warn"
      //#endregion
      // TODO: go over https://typescript-eslint.io/linting/troubleshooting/performance-troubleshooting
    }
  },
  {
    // Don't do type-aware linting for untyped files
    files: ["**/*.js"],
    ...tsEslint.configs.disableTypeChecked
  }
];
export {
  config_default as default
};
