// @arcgis/eslint-config/dist/chunk-G2LC3KBE.js
import {
  makeEslintPlugin
} from "./chunk-FVTAJXPI.js";

// src/plugins/utils/helpers.ts
var isTestFile = (filePath) => filePath.includes("/test") || filePath.includes(".test") || filePath.includes(".spec") || filePath.includes("e2e") || filePath.includes("__") || filePath.includes("/.");

// src/plugins/webgis/utils/creator.ts
import { ESLintUtils } from "@typescript-eslint/utils";
var createRule = ESLintUtils.RuleCreator(
  (rule) => `https://devtopia.esri.com/WebGIS/arcgis-web-components/tree/main/packages/support-packages/eslint-config/src/plugins/webgis/rules/${rule}.ts`
);

// src/plugins/webgis/rules/no-import-outside-src.ts
import { resolve } from "path/posix";
var description = `Imports of files outside the src/ folder are not-portable and likely to break for consumers of this package.`;
var noImportOutsideSrc = createRule({
  name: "no-import-outside-src",
  meta: {
    docs: {
      description,
      defaultLevel: "error"
    },
    messages: {
      noImportOutsideSrc: description
    },
    type: "problem",
    schema: []
  },
  defaultOptions: [],
  create(context) {
    const fileName = context.filename;
    const srcIndex = fileName.indexOf("/src/");
    if (srcIndex === -1) {
      return {};
    }
    const basePath = fileName.slice(0, srcIndex + "/src/".length);
    if (isTestFile(fileName)) {
      return {};
    }
    return {
      ImportDeclaration(node) {
        const specifier = node.source.value;
        const isRelativeAncestor = specifier.startsWith("../");
        if (!isRelativeAncestor) {
          return;
        }
        const relativePath = resolve(fileName, "..", specifier);
        if (!relativePath.startsWith(basePath) && !relativePath.endsWith(".json")) {
          context.report({
            messageId: "noImportOutsideSrc",
            node: node.source
          });
        }
      }
    };
  }
});

// src/plugins/webgis/rules/no-touching-jsdoc.ts
var description2 = `Having two JSDoc comments next to each other is most likely a mistake - consider combining them into one, or separating them for clarity.`;
var noTouchingJsDoc = createRule({
  name: "no-touching-jsdoc",
  meta: {
    docs: {
      description: description2,
      defaultLevel: "warn"
    },
    messages: {
      noTouchingJsDoc: description2
    },
    type: "problem",
    schema: []
  },
  defaultOptions: [],
  create(context) {
    return {
      Program() {
        Array.from(context.sourceCode.text.matchAll(reTouchingJsDoc), (match) => {
          const previousJsDocStart = context.sourceCode.text.slice(0, match.index).lastIndexOf("/**");
          if (previousJsDocStart === -1) {
            return void 0;
          }
          const previousJsDocContent = context.sourceCode.text.slice(
            previousJsDocStart + "/**".length,
            match.index - "*/".length
          );
          const containsOtherComments = previousJsDocContent.includes("*/");
          if (containsOtherComments) {
            return void 0;
          }
          context.report({
            messageId: "noTouchingJsDoc",
            loc: {
              start: context.sourceCode.getLocFromIndex(match.index),
              end: context.sourceCode.getLocFromIndex(match.index + match[0].length)
            }
          });
          return void 0;
        });
      }
    };
  }
});
var reTouchingJsDoc = /\*\/\s+\/\*\*/gu;

// src/plugins/webgis/rules/require-js-in-core-import.ts
import { AST_NODE_TYPES } from "@typescript-eslint/utils";
var description3 = `@arcgis/core imports need to end with .js for better compatibility with @arcgis/core ESM CDN builds.`;
var requireJsInCoreImport = createRule({
  name: "require-js-in-core-import",
  meta: {
    docs: {
      description: description3,
      defaultLevel: "warn"
    },
    messages: {
      requireJsInCoreImport: description3
    },
    type: "problem",
    fixable: "code",
    schema: []
  },
  defaultOptions: [],
  create(context) {
    if (isTestFile(context.filename)) {
      return {};
    }
    function updateSpecifier(node) {
      if (node.source.type !== AST_NODE_TYPES.Literal || node.importKind === "type") {
        return;
      }
      const specifier = node.source.value;
      if (typeof specifier !== "string" || !specifier.startsWith("@arcgis/core/") || // Already ends with .js or .json
      specifier.includes(".")) {
        return;
      }
      context.report({
        node: node.source,
        messageId: "requireJsInCoreImport",
        fix: (fixer) => fixer.replaceText(node.source, `"${specifier}.js"`)
      });
    }
    return {
      ImportDeclaration: updateSpecifier,
      ImportExpression: updateSpecifier
    };
  }
});

// src/plugins/webgis/index.ts
var webgis_default = makeEslintPlugin("webgis", {
  "no-import-outside-src": noImportOutsideSrc,
  "no-touching-jsdoc": noTouchingJsDoc,
  "require-js-in-core-import": requireJsInCoreImport
});

export {
  webgis_default
};
